src/coreclr/System.Private.CoreLib/src/System/MulticastDelegate.cs

```diff
@@ -156,22 +156,22 @@ private bool InvocationListEquals(MulticastDelegate d)
 
         private static bool TrySetSlot(object?[] a, int index, object o)
         {
-            if (a[index] == null && System.Threading.Interlocked.CompareExchange<object?>(ref a[index], o, null) == null)
+            ref object? slot = ref a[index];
+            object? slotVal = slot;
+
+            if (slotVal is null && System.Threading.Interlocked.CompareExchange<object?>(ref slot, o, null) is null)
                 return true;
 
             // The slot may be already set because we have added and removed the same method before.
             // Optimize this case, because it's cheaper than copying the array.
-            if (a[index] is object ai)
-            {
-                MulticastDelegate d = (MulticastDelegate)o;
-                MulticastDelegate dd = (MulticastDelegate)ai;
+            MulticastDelegate d = (MulticastDelegate)o;
+            MulticastDelegate dd = (MulticastDelegate)slotVal!;
```

Comment:
This is a threading behavioral change: it should deref _slot_ again rather than use the cached _slotVal_ value.

Reply:
As `CompareExchange` didn't return null we know the value did not change, so we can avoid reading memory again and use the cached value. However, there is technically a change in the threading behaviour.

In the case of multi-threading, the existing behaviour allows `TrySetSlot` to return false, even though setting the slot "succeeded".  The slot may be written again by some other thread, but that is not the concern of this method.

Reply:
> As CompareExchange didn't return null we know the value did not change, so we can avoid reading memory again and use the cached value.

My concern was that it is possible for the initial "is null?" check to pass but for `CompareExchange` to fail. The old code and the new code have different observable behaviors given this scenario. Hence my comment on the PR.

Are you saying that this scenario either isn't possible in practice, or that there's existing code elsewhere that handles this condition and that it doesn't need to be special-cased here?

Reply:
Right, I see the logic error. We need to keep the return value of `CompareExchange` in case the exchange fails.